cpp/src/arrow/util/decimal.cc

```diff
@@ -0,0 +1,97 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/decimal.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+Status StringToInteger(
+    const std::string& whole, const std::string& fractional, int8_t sign, int32_t* out) {
+  DCHECK(!whole.empty() || !fractional.empty());
+  if (!whole.empty()) {
+    *out = std::stoi(whole, nullptr, 10) *
+           static_cast<int32_t>(pow(10.0, static_cast<double>(fractional.size())));
+  }
+  if (!fractional.empty()) { *out += std::stoi(fractional, nullptr, 10); }
+  *out *= sign;
+  return Status::OK();
+}
+
+Status StringToInteger(
+    const std::string& whole, const std::string& fractional, int8_t sign, int64_t* out) {
+  DCHECK(!whole.empty() || !fractional.empty());
+  if (!whole.empty()) {
+    *out = static_cast<int64_t>(std::stoll(whole, nullptr, 10)) *
+           static_cast<int64_t>(pow(10.0, static_cast<double>(fractional.size())));
+  }
+  if (!fractional.empty()) { *out += std::stoll(fractional, nullptr, 10); }
+  *out *= sign;
+  return Status::OK();
+}
+
+Status StringToInteger(
+    const std::string& whole, const std::string& fractional, int8_t sign, int128_t* out) {
+  DCHECK(!whole.empty() || !fractional.empty());
+  *out = int128_t(whole + fractional) * sign;
+  return Status::OK();
+}
+
+Status FromBytes(const uint8_t* bytes, Decimal32* value) {
+  value->value = *reinterpret_cast<const int32_t*>(bytes);
+  return Status::OK();
+}
+
+Status FromBytes(const uint8_t* bytes, Decimal64* value) {
+  value->value = *reinterpret_cast<const int64_t*>(bytes);
+  return Status::OK();
+}
+
+Status FromBytes(const uint8_t* bytes, bool is_negative, Decimal128* decimal) {
+  auto& decimal_value(decimal->value);
+  int128_t::backend_type& backend(decimal_value.backend());
```

Comment:
i can't do that because i'm mutating the backend with `resize` and `memcpy` below.